flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogDelegateStateTest.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.flink.runtime.state.VoidNamespace;
+import org.apache.flink.runtime.state.VoidNamespaceSerializer;
+import org.apache.flink.runtime.state.hashmap.HashMapStateBackend;
+import org.apache.flink.util.IOUtils;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static org.apache.flink.state.changelog.ChangelogStateBackendTestUtils.createKeyedBackend;
+import static org.junit.Assert.assertSame;
+
+/** Tests for {@link ChangelogStateBackend} delegating state accesses. */
+@RunWith(Parameterized.class)
+public class ChangelogDelegateStateTest {
+    private MockEnvironment env;
+
+    @Parameterized.Parameters
+    public static List<Supplier<AbstractStateBackend>> delegatedStateBackend() {
+        return Arrays.asList(HashMapStateBackend::new, EmbeddedRocksDBStateBackend::new);
+    }
+
+    @Parameterized.Parameter public Supplier<AbstractStateBackend> backend;
+
+    @Before
+    public void before() {
+        env = MockEnvironment.builder().build();
+    }
+
+    @After
+    public void after() {
+        IOUtils.closeQuietly(env);
+    }
+
+    @Test
+    public void testDelegatingValueState() throws Exception {
+        testDelegatingState(
+                new ValueStateDescriptor<>("id", String.class), ChangelogValueState.class);
+    }
+
+    @Test
+    public void testDelegatingListState() throws Exception {
+        testDelegatingState(
+                new ListStateDescriptor<>("id", String.class), ChangelogListState.class);
+    }
+
+    @Test
+    public void testDelegatingMapState() throws Exception {
+        testDelegatingState(
+                new MapStateDescriptor<>("id", Integer.class, String.class),
+                ChangelogMapState.class);
+    }
+
+    @Test
+    public void testDelegatingReducingState() throws Exception {
+        testDelegatingState(
+                new ReducingStateDescriptor<>(
+                        "id", (value1, value2) -> value1 + "," + value2, String.class),
+                ChangelogReducingState.class);
+    }
+
+    @Test
+    public void testDelegatingAggregatingState() throws Exception {
+        testDelegatingState(
+                new AggregatingStateDescriptor<>(
+                        "my-state",
+                        new StateBackendTestBase.MutableAggregatingAddingFunction(),
+                        StateBackendTestBase.MutableLong.class),
+                ChangelogAggregatingState.class);
+    }
+
+    private void testDelegatingState(StateDescriptor descriptor, Class<?> stateClass)
+            throws Exception {
+        KeyedStateBackend<Integer> delegatedBackend = createKeyedBackend(backend.get(), env);
+        KeyedStateBackend<Integer> changelogBackend =
+                createKeyedBackend(new ChangelogStateBackend(backend.get()), env);
```

Comment:
Why we have to create the `delegatedBackend` explicitly? Can we just get the delegated keyed backend from `changelogBackend` directly?

Reply:
That's what we want to compare against.

One is created from the unwrapped state backend, and the other is created from the wrapped state backend to make sure the wrapped one has the right wrapped state + the underlying state class is consistent with the unwrapped one.